Add chapter on template expressions. (#247)
amzn/ion-docs: src/template-expr.adoc

[[sec:templexpr]]
== Template Expressions

include::styles.adoc[]

The behavior of a macro is defined in terms of an expression language.
Like encoding directives and modules, this language is expressed as Ion data, and
the meaning of templates is defined structurally and recursively based on the Ion data model.

=== Grammar

Here's the relevant portion of the <<sec:grammar,domain grammar>>:

include::grammar.adoc[tag=templexpr]

An expression in this language is called a _template_, and the expansion of a template (that is,
its evaluation) produces a stream of Ion values.
The central design concept is that symbols denote variable references, S-expressions denote
operator invocations, and other Ion types denote values of that type.

==== Symbols are Variable References

When a template is an Ion symbol, it denotes a reference to a variable, either a macro parameter

> or a local binding from for.

Reviewer: I had a bit of trouble parsing this sentence:
```suggestion
When a template is an Ion symbol, it denotes a reference to a variable, either a macro parameter
or a local binding from a `for` expression.
```

> The result of this template is the stream of values produced by that variable.

Reviewer: Up for discussion, but I don't think of a variable as 'producing' values.

```suggestion
The result of this template is the stream of values referred to by that variable.
```

We say “quasi” literal because the elements of the container are treated as templates, not
literal values.

When a template is a list, it expands into a list with the same annotations.
The elements of the list-template are each treated as templates themselves.
Each sub-template may produce any number of values, and the resulting streams are all
concatenated to produce the output list.

----
[1, [2, 3], 4]       ⇒ [1, [2, 3], 4]
[1, (values 2 3), 4] ⇒ [1, 2, 3, 4]
[1, (values), 3]     ⇒ [1, 3]
----

When a template is a struct, it expands into a struct with the same annotations.
The struct-template’s field names are treated as literals, and field values are treated as
sub-templates, and the output struct contains the given names and their associated sub-template
expansions.

Field-value sub-templates MAY produce multiple values.
When a sub-template produces more than one result, then the output struct will have more than one
field with the same name.
When a sub-template produces no results, then nothing is added to the output.

----
{a:(values 1 2)} ⇒ {a:1, a:2}   // or, equivalently, {a:2, a:1}
{f:(values)}     ⇒ {}
----

> 

Reviewer: I expected a blurb about how struct splicing works. If you agree, could you add a `TODO`?

Author: I'm not sure what you mean.  I don't think that the template language has a way to splice a struct into a containing struct, because there's no way to omit the field-name from the syntax.

If you are referring to the "fanout" on L73, that's explained on L68.

The first element of the S-expression must be a symbol that identifies the operator, and the
meaning of subsequent elements depends on the operator.

Operators come in two varieties: special forms and macro invocations.

=== Special Forms

Special forms are operators that cannot be expressed as macros, because some parts of their
syntax are not recursively-expanded templates, as all macro arguments are.

We use `*bold monospace*` when naming these special forms, to distinguish them from macro names.

In the descriptions below, `_template_` subforms accept any template-language form.
In all such cases, sub-templates are expanded only when indicated.

==== Preventing Evaluation

===== `*literal*`

[{nrm}]
----
(*literal* _datum_)
----

Produces _datum_ as-is, preventing the operand from being evaluated as a template.

For example, `(*literal* [1, (values 2 3), 4])` produces `[1, (values 2 3), 4]`; both the list
and the S-expression are treated as literal, constant data, not as template expressions to be
expanded.

==== Conditionals

These special forms allow output to vary based on whether a template produces zero, one, or more
values.

===== `*if_void*`

[{nrm}]
----

> (*if_void* _template1_ _template2_ _template3_)

Reviewer: The current names are ok, but I wonder if we could use something more descriptive.

```suggestion
(*if_void* _target_template_ _true_template_ _false_template_)
```

Author: I've updated this to use descriptive subscripts.

----

> Evaluates templates conditionally based on cardinality of a stream.

Reviewer: ```suggestion
Evaluates templates conditionally based on the cardinality of a stream.
```

(:decimal_constraint (1 5) (-5 0)) ⇒ { precision: range::[1, 5],
                                       exponent: range::[-5, 0] }
(:decimal_constraint (:) (3 max))  ⇒ { exponent: range::[3, max] }
(:decimal_constraint (1) (:))      ⇒ { precision: 1 }
----

==== Mapping

These special forms produce repeated output mapped across elements of a stream.

===== `*for*`

[{nrm}]
----
(*for* [(_id_ _in-template_), ...] _body-template_)
----

Iteratively expands the _body-template_ using individual values from the _in-templates_.

Each iteration takes the next value from each _in-template_ stream; iteration stops when any
stream ends.
Local variables are created for each identifier _id_, bound to the current value from their stream.
The _body-template_ is then expanded in that environment, and iteration proceeds.
The result of the `*for*` expression is the concatenated results of the body expansions.

NOTE: The termination rule is under discussion; see
https://github.com/amazon-ion/ion-docs/issues/201

=== Macro Invocation

A macro definition can express its output in terms of other macros. Quite often, these will be
macros provided by the Ion implementation,

> // —everything in System Macros 2023-05 is available by default—

Reviewer: I think this is an internal doc leftover.

Author: It is, but I want to resurrect it with a proper link (the TODO on L205)

When a template is an S-expression and the first element is not the name of a special form, that
element must instead be a _macro-ref_ and the template denotes a macro invocation.
There are multiple sources of macros: the defining module’s internal environment (which is being
incrementally extended with each definition), and the exported macros of modules loaded by the
enclosing module or `$ion_encoding` directive.

// TODO See Resolving Macro References: Encoding Modules 2023-05 for the relevant algorithm.

The remaining elements of the S-expression are subforms that denote the inputs to the macro.
These use normal Ion notation, but what’s syntactically acceptable is defined by the macro’s
signature.

The number of such subforms (that is, the invocation’s actual arity) must be equal to or greater
than the macro’s minimum arity, and at most its maximum arity, when one exists.
In other words, an invocation must contain one subform for each required parameter, followed by
optional subforms for the remaining optional parameters.

Within an invocation expression, the syntax of each subform is defined first by its parameter’s

> grouping form, then its base type:

Reviewer: I think some examples would make this section easier for folks to digest without needing to refer to the glossary. (I'm fine with a `TODO`.)

Author: Added TODO
